src/HasManyDeep.php

```diff
@@ -11,6 +11,10 @@
 use Illuminate\Pagination\CursorPaginator;
 use Illuminate\Support\Collection;
 
+/**
+ * @template TRelatedModel of \Illuminate\Database\Eloquent\Model
+ * @extends Relation<TRelatedModel>
```

Comment:
Can you fully-qualify the class name? At the moment, PhpStorm warns about "Undefined class 'Relation'". In the linked example, the `Relation` class is already imported.

```suggestion
 * @extends \Illuminate\Database\Eloquent\Relations\Relation<TRelatedModel>
```